### ISSUE TITLE
Android O compatibility

### ISSUE THREAD

**Issue by danielfaust** [mendhak/gpslogger#560]
Opening this issue for Android O related discussion. I am not sure if this is a good place for it, but it could be.

I installed Android O yesterday, and took it two times for a bike ride since then. Yesterday I forgot to enable recording to kml, so I can't say anything about that ride, but the CSV does look good.

Today's ride logged all without any problems whatsoever. I had the app in foreground with the screen on, still need to see how it behaves when it's running in the background.

Oh, yesterday I had an issue where the "Start recording" button didn't react at all, it didn't even get depressed. Turning the screen off and on again fixed the issue. Maybe it was the bike mount pressing on the volume button, not sure though.

**Comment by mendhak** [comment]
What phone did you install O on? I know there are [preview images](https://developer.android.com/preview/download.html) available which I could probably flash to some phones. 

Background test would be useful!  I know O is aggressive from their release notes so I wonder what kind of effect it will have.

**Comment by danielfaust** [comment]
Nexus 5X.

I have another app which is a background service that collects sensor data every minute. This one has stopped working a couple of times for some minutes (up to 30) since the upgrade, mostly when the screen was off. Like 8 times in 24h, two from those while the screen was on. This did not happen before.

During both bike rides it remained active. I will check it nonetheless by putting GPSLogger in the background. But I have no comparison to it, since on Nougat I always used in the foreground as well.

**Comment by danielfaust** [comment]
I placed the phone r/outside and left it logging on a bench.

One hour it was in the foreground, the second hour in the background, where I manually placed it by starting another app.

In the first hour, after 30 mins the screen went off, that is my display timeout.

I've placed annotations in the csv file.

Logging is set to every two seconds.

[gps--2017-06-12--mon--19-25--88--android-o.csv.txt](https://github.com/mendhak/gpslogger/files/1069332/gps--2017-06-12--mon--19-25--88--android-o.csv.txt)

**Comment by mendhak** [comment]
I had a look at your file and it seems to be mostly near the logging interval, with one >1 minute gap towards the end.  

I had different results.  I ran it for 16 hours but with default 60s value and got several deep-doze gaps of >1 hour.  Most of these corresponded to me moving occasionally.   I'll try out a few more settings including your 2s and try to move about a bit more (depends on work...)

Also running on a Nexus 5X with Android O Preview image. 

I'll be slow but I'll be adding some test results to a spreadsheet as I go:

https://docs.google.com/spreadsheets/d/17n6v-aa9OkdpVqGn1k9yYIV4XN0VF6y0kFjqBG7y3gM/edit?usp=sharing

Also - thanks for starting this thread, it's a good proactive idea.

**Comment by danielfaust** [comment]
I'm having really big problems with my background sensor logging app, which is supposed to run 24/7. Android O is going to break a lot of apps on the market.

Initially it wasn't that noticeable, but I now can't rely on it anymore. It gets shut down so often and can't restart itself because the Alarm Broadcast is non-existent (need to reprogram it). I'm annoyed.

Expect to run into problems with GPSLogger with long, multihour runs. Mine are no longer than 40 minutes with 2 sec intervals and it's working like a charm, even in background and with the screen turned off.

**Comment by mendhak** [comment]
I know, I feel like with every subsequent release, all apps are being punished because a few apps are misbehaving. But then if this isn't an API representation of human society as a disjointed collective, I don't know what is. :thinking:

**Comment by danielfaust** [comment]
This is a "censored" log of the service running during the last 20 days. 20 days ago is at the top, today at the bottom. The red arrow indicates when I upgraded to Android O. The dark bars indicate where data got effectively lost because the service got shut down. Sometimes the service gets started automatically by the OS, I don't know what the criteria for that is, usually I need to start it manually.

![20 days](https://user-images.githubusercontent.com/320098/27407804-369c3394-56da-11e7-8d27-e8497747fb56.png)

My phone isn't snappier, the battery is just as good/bad as always and I have no means of telling the OS that this is a high-priority app for me and I always want it to be active no matter what.

**Comment by ithinkihaveacat** [comment]
I don't know if the app already does this, but a foreground service that displays a persistent [ongoing notification](https://developer.android.com/guide/components/services.html#Foreground) gets access to more frequent location scans in Android O ([more info](https://developer.android.com/preview/features/background-location-limits.html)).

**Comment by danielfaust** [comment]
Turns out that it was an exception caused by trying to list the foreground apps. That worked/not-worked randomly... try-catch solved it.

I noticed a couple of weeks ago that there is a setting in
Settings -> Apps -> Advanced -> Special App-Access -> Battery-Optimization -> There you select the app and choose "Not Optimized" (my phone is set to German, so I'm translating here). Not all apps are listed by default, you have to change the filter from "Not optimized" to "All".

GPSLogger shows as optimized. When I touch it, the popup let's me choose between "Optimize - recommended for longer battery usage" and "Do not optimize - Battery will probably drain faster". I have GPSLogger as optimized, but have no problems with that setting. Since you mentioned that you have seen problems in that >16h run, you could test if removing the optimization helps and eventually put a notice in your docs.

I don't know if that setting already existed in Android 7.0, I don't have that version on any device to check.

**Comment by mendhak** [comment]
@ithinkihaveacat yep, it's a persistent ongoing notification to make it a 'foreground' background service.

@danielfaust that is a good point, put it somewhere for users to see.  I read somewhere that Google says I shouldn't prompt the user to unoptimize the app, it as it's considered abusive.
